### PR TITLE
Prediction

### PR DESCRIPTION
**Pull request by samwang141224**
Codes for using the model to detect DLAs

### CODE REVIEW

File: desidlas/datasets/DesiMock.py [hunk 1]
```diff
@@ -125,19 +125,39 @@ def get_data(start_point=0, end_point=self.data_size):
...
             overlap_error_r = sightline.error[self.split_point_br:rrgbound+1]
-            test = overlap_error_b<overlap_error_r
-            overlap_flux = np.array([overlap_flux_b[i] if test[i] else overlap_flux_r[i] for i in range(-blfbound+self.split_point_br)])
-            overlap_error = np.array([overlap_error_b[i] if test[i] else overlap_error_r[i] for i in range(-blfbound+self.split_point_br)])
-            sightline.flux = np.hstack((sightline.flux[:blfbound],overlap_flux,sightline.flux[rrgbound:]))
-            sightline.error = np.hstack((sightline.error[:blfbound],overlap_error,sightline.error[rrgbound:]))
-            sightline.loglam = np.hstack((sightline.loglam[:blfbound],sightline.loglam[blfbound:self.split_point_br],sightline.loglam[rrgbound:]))
+            
+            test_1 = overlap_error_b<overlap_error_r
+            overlap_flux_1 = np.array([overlap_flux_b[i] if test_1[i] else overlap_flux_r[i] for i in range(-blfbound+self.split_point_br)])
```
Comment (profxj) [inline desidlas/datasets/DesiMock.py:140]: add some inline docs

File: desidlas/prediction/analyze_prediction.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
+
```
Comment (profxj) [inline desidlas/prediction/analyze_prediction.py]: add a line describing this module

File: desidlas/prediction/analyze_prediction.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    dla_tbl = Table(names=('TARGET_RA','TARGET_DEC', 'ZQSO','Z','TARGETID','S/N','DLAID','NHI','DLA_CONFIDENCE','NHI_STD','ABSORBER_TYPE'),dtype=('float','float','float','float','int','float','str','float','float','float','str'),meta={'EXTNAME': 'DLACAT'})
+    for jj in range(0,len(sightline.prediction.peaks_ixs)):
+        peak=sightline.prediction.peaks_ixs[jj]
+        peak_lam_spectrum = lam_analyse[peak]
+        z_dla = float(peak_lam_spectrum) / 1215.67 - 1
+        peak_lam_rest=float(peak_lam_spectrum)/(1+sightline.z_qso)
+        _, mean_col_density_prediction, std_col_density_prediction, bias_correction =             sightline.prediction.get_coldensity_for_peak(peak)
+
+        absorber_type =  "DLA" if mean_col_density_prediction >= 20.3 else "LYB" if sightline.is_lyb(peak) else "SUBDLA"
+        dla_tbl.add_row((sightline.ra,sightline.dec,sightline.z_qso,float(z_dla),sightline.id,sightline.s2n,str(sightline.id)+'00'+str(jj),float(mean_col_density_prediction),min(1.0,float(sightline.prediction.offset_conv_sum[peak])),float(std_col_density_prediction),absorber_type))
```
Comment (profxj) [inline desidlas/prediction/analyze_prediction.py]: can you make this line easier to read?

File: desidlas/prediction/dla_catalog.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
+from astropy.table import Table, vstack
```
Comment (profxj) [inline desidlas/prediction/dla_catalog.py:2]: add a line describing this module

File: desidlas/prediction/dla_catalog.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    Return
+    ----------------
+    real_dla_tbl: astropy.table.Table object
+    """
+    real_dla_tbl = Table(names=('TARGET_RA','TARGET_DEC', 'ZQSO','Z','TARGETID','S/N','DLAID','NHI','DLA_CONFIDENCE','NHI_STD','ABSORBER_TYPE'),dtype=('float','float','float','float','int','float','str','float','float','float','str'),meta={'EXTNAME': 'DLACAT'})
+    for ii in range(0,len(sightlines)):
+        sightline=sightlines[ii]
+        for dla in sightline.dlas:
+            absorber_type =  "DLA" if dla.col_density >= 20.3 else "SUBDLA"
+            real_dla_tbl.add_row((sightline.ra,sightline.dec,sightline.z_qso,float(dla.central_wavelength/1215.67-1),sightline.id,sightline.s2n,str(sightline.id)+dla.id,float(dla.col_density),1.0,0.0,absorber_type))
```
Comment (profxj) [inline desidlas/prediction/dla_catalog.py]: can you make this easier to read?

File: desidlas/prediction/dla_catalog.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    return real_dla_tbl
+    
+def catalog_fits(sightlines,dlafile=None,qsofile=None):
+    """
+    save real DLA and QSO catalog.
+    
+    Parameters
+    ----------------
+    sightlines: list of 'dla_cnn.data_model.Sightline.Sightline` object
+    dlafile: str
```
Comment (profxj) [inline desidlas/prediction/dla_catalog.py]: explain these variables

File: desidlas/prediction/get_partprediction.py [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+          n_samples, BATCH_SIZE, timeit.default_timer() - timer))
+
+    return pred, conf, offset, coldensity #return four labels
+
+
+
+
+
+
+if __name__ == '__main__':
```
Comment (profxj) [inline desidlas/prediction/get_partprediction.py:92]: is this for testing?  If so, maybe move to tests/

File: desidlas/prediction/pred_sightline.py [hunk 8]
```diff
@@ -0,0 +1,156 @@
+import numpy as np
```
Comment (profxj) [inline desidlas/prediction/pred_sightline.py:2]: add a line describing this module

File: desidlas/prediction/pred_sightline.py [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+    ----------------
+    tp_preds: list
+    fn_num: int
+    fp_num: int
+   
+    """
+    tp_pred=[]
+    fn_num=0
+    fp_num=0
+    #if pred_catalog['label'][0]=='str':
```
Comment (profxj) [inline desidlas/prediction/pred_sightline.py]: remove

File: desidlas/prediction/pred_sightline.py [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+       #pred_catalog.remove_column('label')
+    pred_catalog.add_column('str',name='label')
+    pred_catalog.add_index('DLAID')
+    #if real_catalog['label'][0]=='str':
+       #real_catalog.remove_column('label')
+    real_catalog.add_column('str',name='label')
+    for real_dla in real_catalog:
```
Comment (profxj) [inline desidlas/prediction/pred_sightline.py:68]: add some inline comments throughout

**Review by profxj — CHANGES_REQUESTED** [review]
This is great.

Here are a few additional requests:

- Update `installing.rst`.  Include notes on how to download the models
- Add a new file to `docs/` describing how to run the code.  On mocks and/or real data
- How was the code tested?  I guess the models are too large to use CI..
- Let's get the documentation on `readthedocs`.  I can help

File: desidlas/datasets/DesiMock.py (hunk 1 above)
Comment (zoujiaqi99) [reply desidlas/datasets/DesiMock.py:140]: done

File: desidlas/prediction/analyze_prediction.py (hunk 2 above)
Comment (zoujiaqi99) [reply desidlas/prediction/analyze_prediction.py]: done

File: desidlas/prediction/dla_catalog.py (hunk 4 above)
Comment (zoujiaqi99) [reply desidlas/prediction/dla_catalog.py:2]: done

File: desidlas/prediction/dla_catalog.py (hunk 5 above)
Comment (zoujiaqi99) [reply desidlas/prediction/dla_catalog.py]: done

File: desidlas/prediction/analyze_prediction.py (hunk 3 above)
Comment (zoujiaqi99) [reply desidlas/prediction/analyze_prediction.py]: done

File: desidlas/prediction/dla_catalog.py (hunk 6 above)
Comment (zoujiaqi99) [reply desidlas/prediction/dla_catalog.py]: done

File: desidlas/prediction/pred_sightline.py (hunk 8 above)
Comment (zoujiaqi99) [reply desidlas/prediction/pred_sightline.py:2]: done

File: desidlas/prediction/pred_sightline.py (hunk 9 above)
Comment (zoujiaqi99) [reply desidlas/prediction/pred_sightline.py]: removed these comments

File: desidlas/prediction/pred_sightline.py (hunk 10 above)
Comment (zoujiaqi99) [reply desidlas/prediction/pred_sightline.py:68]: done

File: desidlas/prediction/get_partprediction.py (hunk 7 above)
Comment (samwang141224) [reply desidlas/prediction/get_partprediction.py:92]: This is not for testing. I just want to include all the codes in this py file. So people only need to run ' python get_partprediction.py -xx' to get the prediction result, I have described this in desidlas/notebook/training_prediction.ipynb

**Comment by profxj** [comment]
Let me know when the additional requests are addressed
and I'll approve.  Thx!

**Comment by samwang141224** [comment]
Hi X
We have updated the installing.rst in docs/ to add the download path of our model files.
We have uploaded two notebooks in docs/notebook/ to describe how to run the codes.
The training and prediction codes are not tested yet. As you said, these codes need to run the CNN model. The model files can not be uploaded on the github because they are too large. So I think this is very difficult if we want to test it online.
For your last comments 'build a read the docs' , do you mean that we build an instruction webpage like the pypeit : https://pypeit.readthedocs.io/en/latest/step-by-step.html ? I may need some time to build this because I have never used readthedocs before. Maybe I can firstly build a draft version and you help to update it?
Thanks.

**Comment by profxj** [comment]
We agreed that you'd generate a few tests on prediction and training 
that *use* the model and that we run ourselves but not online.

We'll save `readthedocs` for a future pull request.  Thx!

**Comment by samwang141224** [comment]
> We agreed that you'd generate a few tests on prediction and training
> that _use_ the model and that we run ourselves but not online.
> 
> We'll save `readthedocs` for a future pull request. Thx!

Hi X, I have updated the test files. The code to test the training is  desidlas/training/test_training.py , the code to test the prediction is desidlas/preiction/test_prediction.py . I don't put them in desidla/tests because I am worried that if I put these codes there, these codes will be run when the github check pull request automatically. This may cause a error. But I am not sure if the directory of these two test codes now is proper. Thanks.

File: desidlas/prediction/test_prediction.py [hunk 11]
```diff
@@ -0,0 +1,45 @@
+import numpy as np
```
Comment (profxj) [inline desidlas/prediction/test_prediction.py]: Please put this in `tests/`

Put it into a method, e.g. `def test_prediction()`
and decorate it with something like:

```
remote_data = pytest.mark.skipif(os.getenv('CNN_MODEL') is None,
                                 reason='test requires CNN model')
```

and put this before the method:

`@remote_data`

See this example:

https://github.com/FRBs/FRB/blob/main/frb/tests/test_frbsurveys.py

Comment (profxj) [reply desidlas/prediction/test_prediction.py]: Then you should put the model in a folder and have the 
enrvionmental variable `CNN_MODEL` hold it.  Something like that.

Let me know if you have any questions.

Comment (samwang141224) [reply desidlas/prediction/test_prediction.py]: Hi X, I have put the codes in test/ and add the 'remote_data' module as you suggested. I add a reminder to export the environmental variable in installing.rst  file. The automatic test seems successful now. Please check this PR. Thanks!

**Review by profxj — APPROVED** [review]
great stuff!

**Pull request merged**